Basic block sections

gretay-js opened the pull request:
Allow some basic blocks to be emitted into separate sections, so that they can be reordered, for example, put all cold blocks of a hot function out of the way. 

Main changes are:
- Linear.Llabel records the target section, if any
- Cfg_with_layout records the target section of a block, if any
- Cfg_to_linear emits jumps between sections
- Emit emit section names and extra jumps between sections (that aren't visible in the cfg) instead of implicit fallthrough 

The assignment of blocks to sections is detetrmined separately, not in this PR. 

Should be easier to review commit by commit.

#### TODO

- [ ] more testing

xclerc commented:
I don't know whether layouts should be considered
equal if they differ only on sections, but I think the
`Cfg_equivalence` module should be updated. Either
to state that sections are ignored, or to take them
into account.

gretay-js commented:
> I don't know whether layouts should be considered
> equal if they differ only on sections, but I think the
> `Cfg_equivalence` module should be updated. Either
> to state that sections are ignored, or to take them
> into account.

It would depend on when the Cfg_equivalence check is called, but 
it is not called anywhere at the moment, so I just added a CR. I also rebased 
this PR, and it is ready for your review @xclerc.

mshinwell commented on driver/flambda_backend_args.ml:
```diff
@@ -868,6 +886,11 @@ module Extra_params = struct
+    | "basic-block-sections" ->
+      let options = [ Flambda_backend_flags.basic_block_sections;
+                      Flambda_backend_flags.use_ocamlcfg ] in
+      Compenv.setter ppf (fun b -> b) name options v;
```
Does this disable ocamlcfg if you disable basic block sections?  Not sure that's right.

gretay-js commented on backend/amd64/emit.mlp:
```diff
@@ -201,7 +201,56 @@ let label s = sym (emit_label s)
+(* Name of current function *)
+let function_name = ref ""
+
+(* Keep current block label for -basic-block-sections to get back to it
```
I tried to improve the comment. We need to keep track of the current basic block section name. We may need to emit to .data section  (e.g., jump table) and then need to emit .text.* directive to get back to the current basic block section.
